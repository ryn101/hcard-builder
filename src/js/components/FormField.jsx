import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

/**
 * Component that acts as a wrapper for the creation of a label and it's associated input field.
 * This component has two forms:
 *
 * - Shorthand - Allows the creation of the component without the need to specify a
 *               input as a child as this is auto generated. When using this form,
 *               the two supported properties of `onFieldChange` and `type` will be applied
 *               directly on the input.
 * - Nested - Used for finer attribute control of the input. This requires an explicit
 *            declaration of an input as a child of this component.
 *
 * In both forms, the fieldId will be applied to the input as it's id.
 */
function FormField({
    fieldId, as, label, children, type, onFieldChange, className,
}) {
    const FormFieldTag = as;

    return (
        <FormFieldTag className={classNames('form__field', className)}>
            {
                // Optionally render the label if the property is provided
                label ? <label className="form__label" htmlFor={fieldId}>{label}</label> : ''
            }
            {
                children
                    // If a child input element is provided, then set the field id
                    ? React.cloneElement(children, { id: fieldId })
                    : <input id={fieldId} type={type} onChange={onFieldChange} />
            }
        </FormFieldTag>
    );
}

FormField.propTypes = {
    fieldId: PropTypes.string.isRequired,
    as: PropTypes.oneOf(['div', 'li', 'p']),
    label: PropTypes.string,
    children: PropTypes.element,
    type: PropTypes.string,
    onFieldChange: PropTypes.func,
    className: PropTypes.string,
};

FormField.defaultProps = {
    as: 'div',
    label: '',
    children: null,
    type: 'text',
    onFieldChange: () => {},
    className: '',
};

export default FormField;
