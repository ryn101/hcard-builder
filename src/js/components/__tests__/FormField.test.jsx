import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import FormField from '../FormField';

configure({ adapter: new Adapter() });

const executeSuiteForJsxLiteral = (jsxLiteralCreator) => {
    let formFieldNode;
    const label = 'The input label description';
    const fieldId = 'formInput';
    const onFieldChange = jest.fn();

    beforeEach(() => {
        formFieldNode = shallow(jsxLiteralCreator(label, fieldId, onFieldChange));
    });

    it('renders with state correctly', () => {
        expect(formFieldNode).toMatchSnapshot();
    });

    it('renders the label correctly', () => {
        const labelSelector = '.form__label';
        const labelNode = formFieldNode.find(labelSelector);
        expect(labelNode.exists).toBeTruthy();
        expect(labelNode.text()).toEqual(label);

        formFieldNode.setProps({ label: null });
        expect(formFieldNode.find(labelSelector).exists()).toBeFalsy();
    });

    it('renders the children correctly', () => {
        const childNodes = formFieldNode.children();
        expect(childNodes).toHaveLength(2);

        const formInput = childNodes.find('input');
        expect(formInput.exists()).toBeTruthy();
        expect(formInput.prop('id')).toEqual(fieldId);
    });

    it('calls the on field update method when updating an input', () => {
        const formInput = formFieldNode.find('input');
        const changeEvent = { target: { id: fieldId, value: 'new value' } };
        formInput.simulate('change', changeEvent);
        expect(onFieldChange).toBeCalledTimes(1);
        expect(onFieldChange).toBeCalledWith(changeEvent);
    });
};

it('renders the form input group without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
        <FormField label="Title" fieldId="formInput" />,
        div,
    );

    ReactDOM.unmountComponentAtNode(div);
});

describe('once rendered shorthand', () => {
    executeSuiteForJsxLiteral((label, fieldId, onFieldChange) => (
        <FormField label={label} fieldId={fieldId} onFieldChange={onFieldChange} />
    ));
});

describe('once rendered', () => {
    executeSuiteForJsxLiteral((label, fieldId, onFieldChange) => (
        <FormField label={label} fieldId={fieldId}>
            <input onChange={onFieldChange} />
        </FormField>
    ));
});
