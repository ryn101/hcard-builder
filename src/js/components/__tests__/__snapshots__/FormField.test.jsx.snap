// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`once rendered renders with state correctly 1`] = `
<div
  className="form__field"
>
  <label
    className="form__label"
    htmlFor="formInput"
  >
    The input label description
  </label>
  <input
    id="formInput"
    onChange={[MockFunction]}
  />
</div>
`;

exports[`once rendered shorthand renders with state correctly 1`] = `
<div
  className="form__field"
>
  <label
    className="form__label"
    htmlFor="formInput"
  >
    The input label description
  </label>
  <input
    id="formInput"
    onChange={[MockFunction]}
    type="text"
  />
</div>
`;
